akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -812,13 +812,20 @@ private[remote] abstract class ArteryTransport(_system: ExtendedActorSystem, _pr
   // Checks for termination hint messages and sends an ACK for those (not processing them further)
   // Purpose of this stage is flushing, the sender can wait for the ACKs up to try flushing
   // pending messages.
-  def terminationHintReplier(): Flow[InboundEnvelope, InboundEnvelope, NotUsed] = {
+  def terminationHintReplier(inControlStream: Boolean): Flow[InboundEnvelope, InboundEnvelope, NotUsed] = {
     Flow[InboundEnvelope].filter { envelope ⇒
       envelope.message match {
-        case _: ActorSystemTerminating ⇒
+        case ActorSystemTerminating(from) ⇒
           envelope.sender match {
-            case OptionVal.Some(snd) ⇒ snd.tell(ActorSystemTerminatingAck(localAddress), ActorRef.noSender)
-            case OptionVal.None      ⇒ log.error("Expected sender for ActorSystemTerminating message")
+            case OptionVal.Some(snd) ⇒
+              snd.tell(ActorSystemTerminatingAck(localAddress), ActorRef.noSender)
+              if (inControlStream)
+                system.scheduler.scheduleOnce(settings.Advanced.ShutdownFlushTimeout) {
+                  if (!isShutdown)
+                    quarantine(from.address, Some(from.uid), "ActorSystem terminated")
```

Comment:
This improves the situation for the normal ActorSystem shutdown since we cleanup many things when it's quarantined. It will not capture the case when the remote `kill -9`.

For a cluster the situation much better because we quarantine removed members.

Reply:
I have confirmed that this is also solving the leak of `ActorRef`s that was reported in https://github.com/akka/akka/issues/24973

I'll look into the `kill -9` case for that now.

Reply:
To clarify, some `ActorRef` remains in the `ActorRefResolveCache`, but its size is bounded to 1024. I have verified that this bound works in practice by setting it to a lower value.